text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,1001 @@
+was read or referenced from. Consequently, what safety proofs one must write when using such a
+wrapper depend on the dataflow of the program.
+
+And worse, certain dangerous flows do not require `unsafe` at all. For instance, unsafe fields of
+the same type can be laundered between fields with different invariants; safe code could exchange
+`Even` and `Odd`s' `val`s:
+
+```rust
+struct Even {
+    val: Unsafe<usize>,
+}
+
+struct Odd {
+    val: Unsafe<usize>,
+}
+```
+
+We can plug this particular hole by adding a type parameter to `Unsafe` that encodes the type of the
+outer datatype, `O`; e.g.:
+
+```rust
+#[repr(transparent)]
+pub struct Unsafe<O: ?Sized, T: ?Sized>(PhantomData<O>, T);
+```
+
+However, it remains possible to exchange unsafe fields within the same type; for example, safe code
+can freely exchange the values of `len` and `cap` of this hypothetical vector:
+
+```rust
+struct Vec<T> {
+    alloc: Unsafe<Self, *mut T>,
+    len: Unsafe<Self, usize>,
+    cap: Unsafe<Self, usize>,
+}
+```
+
+The [`unsafe-fields`](https://crates.io/crates/unsafe-fields) crate plugs this hole by extending
+`Unsafe` with a const generic that holds a hash of the field name. Even so, it remains possible for
+safe code to exchange the same unsafe field between different instances of the same type (e.g.,
+exchanging the `len`s of two instances of the aforementioned `Vec`).
+
+These challenges motivate first-class support for field safety tooling.
+
+### More Syntactic Granularity
+
+This RFC proposes the rule that *a field marked `unsafe` is unsafe to use*. This rule is flexible
+enough to handle arbitrary field invariants, but — in some scenarios — requires that the user write
+trivial safety comments. For example, in some scenarios, an unsafe is trivially sound to read:
+
+```rust
+struct Even {
+    /// # Safety
+    ///
+    /// `val` is an even number.
+    val: u8,
+}
+
+impl Into<u8> for Even {
+    fn into(self) -> u8 {
+        // SAFETY: Reading this `val` cannot
+        // violate its invariant.
+        unsafe { self.val }
+    }
+}
+```
+
+In other scenarios, an unsafe field is trivially sound to `&`-reference (but not `&mut`-reference).
+
+Since it is impossible for the compiler to precisely determine the safety requirements of an unsafe
+field from a type-directed analysis, we must *either* choose a usage rule that fits all scenarios
+(i.e., the approach adopted by this RFC) *or* provide the user with a mechanism to signal their
+requirements to the compiler. Here, we explore this alternative.
+
+The design space of syntactic knobs is vast. For instance, we could require that the user enumerate
+the operations that require `unsafe`; e.g.:
+
+- `unsafe(init,&mut,&,read)` (everything is unsafe)
+- `unsafe(init,&mut,&)` (everything except reading unsafe)
+- `unsafe(init,&mut)` (everything except reading and `&`-referencing unsafe)
+- etc.
+
+Besides the unclear semantics of an unparameterized `unsafe()`, this design has the disadvantage
```

Comment:
The way I conceptualized it in my proposal at https://github.com/rust-lang/rfcs/pull/3458#discussion_r2040267818: the argument to `unsafe` denotes the weakest reference type that is `unsafe` to obtain. So `unsafe()` makes `&` and `&mut` references unsafe, while `unsafe(mut)` only makes `&mut` unsafe.

Reply:
If Rust ever gets partial borrows, the syntax for that could be adopted to extend this feature as well.